keps/sig-storage/1710-selinux-relabeling/README.md

```diff
@@ -51,223 +48,183 @@
-Therefore rogue user cannot access potentially secret data of other containers, because volumes of each container have different label.
+Only process with the context `...:container_t:s0:c309,c383` can access files with label `container_file_t:s0:c309,c383`.
+Therefore, a rogue user who escaped boundaries of its container cannot access data of other containers, because volumes of each container have different label.
+Even processes running as root (UID 0) are denied access to these files, unless they run with the right SELinux context. 
 
-In further text, we're going to shorten both `system_u:system_r:container_t:s0:c309,c383` (context of a process) and `system_u:object_r:container_file_t:s0:c309,c383` (label of a file) to `s0:309:383`.
+Further in this KEP we assume that the SELinux is enabled on the system. This KEP has absolutely no effect on systems that run without SELinux. Kubelet already knows if SELinux is enabled and does not do anything with it if it's disabled or not available (e.g. on Windows). 
 
 See [SELinux documentation](https://selinuxproject.org/page/NB_MLS) for more details.
 
+
 ### SELinux context assignment
 In Kubernetes, the SELinux context of a pod is assigned in two ways:
 1. Either it is set by user in PodSpec or Container: https://kubernetes.io/docs/tasks/configure-pod-container/security-context/.
 1. If not set in Pod/Container, the container runtime will allocate a new unique SELinux context and assign it to a pod (container) by itself.
 
-### Volumes
-Currently, Kubernetes *knows*(*) which volume plugins supports SELinux (i.e. supports extended attributes on a filesystem the plugin provides).
-If SELinux is supported for a volume, it passes the volume to the container runtime with ":Z" option ("private unshared").
-The container runtime then **recursively relabels** all files on the volume to either the label set in PodSpec/Container or the random value allocated by the container runtime itself.
+### Volume mounting
+Linux kernel, with SELinux compiled in, allows `mount -o context=system_u:system_r:container_t:s0:c309,c383 <what> <where>` to mount a volume and pretend that all files on the volume have given SELinux label.
+
+Following conditions must be met:
+* It must be the first mount of the volume! It does not work for subsequent mounts of the volume, incl. bind-mounts.
+    * Second mount of the same volume fails with `mount point not mounted or bad option` when the `-o context=` does not match the first mount.
+    * A bind mount inherits `-o context=XYZ` option from the original mount (if it was set there). 
+* `/bin/mount` process must detect that the kernel supports SELinux. [It does so](https://github.com/util-linux/util-linux/blob/441f9b9303d015f1777aec7168807d58feacca31/libmount/src/context_mount.c#L291) by checking that both `/etc/selinux/config` file exists and `/sys/fs/selinux` is a mount point with "selinuxfs" type.
+    * `/bind/mount` throws away the SELinux mount options otherwise.
```

Comment:
There is no indication that `/bin/mount` drops SELinux mount options. See util-linux code. It's quite unfortunate.

> Do we fallback to old behaviour or try to run pod without any selinux labels (via mount or otherwise?)

That's a good question. For CSI, the driver *did* advertise that it supports `-o context`. so it's driver's fault if it did not apply the context. We can't check mount options in kubelet, because not all CSI drivers mount or they may mount in a weird way. We could check context of the root of the volume and fail, if it has unexpected context.

Reply:
>We could check context of the root of the volume and fail, if it has unexpected context.

I like this option if we can manage it. Can we add this detail to the KEP?